dashboard/app/controllers/projects_controller.rb

```diff
@@ -310,19 +310,37 @@ def create_new
     )
   end
 
-  # GET /projects(/script/:script_id)/level/:level_id
+  # GET /projects(/script/:script_id)/level/:level_id(/user/:user_id)
   # Given a level_id and the current user (or signed out user), get the existing project
   # or create a new project for that level and user. If a script_id is provided, get or
   # create the project for that level, script and user
   # Returns json: {channel: <encrypted-channel-token>}
   def get_or_create_for_level
     script_id = params[:script_id]
+    user_id = params[:user_id]
     level = Level.find(params[:level_id])
+
+    # If viewing another user's work, ensure that we have permission.
+    if user_id
+      script_level = level.script_levels.find_by_script_id(script_id)
+      user = User.find(user_id)
+      unless can?(:view_as_user, script_level, user)
```

Comment:
This is based on: https://github.com/code-dot-org/code-dot-org/blob/f280b95523fb6a3aabbe624250146256a9480cef/dashboard/app/controllers/api_controller.rb#L479